File: src/machine/rollingEngine.ts
/**
 * ENGINEERING CALCULATIONS LAYER — the "Engineering Calculations" box in the §4
 * architecture diagram.
 *
 * Derived engineering indicators computed FROM MachineState for the health and
 * KPI panels. Everything here is a pure function of state: a utilisation
 * percentage, a ratio, a consistency check. No new process values are invented,
 * and nothing here feeds back into the simulation.
 *
 * These live in one module rather than in the components that display them so
 * that "force utilisation" means exactly one thing everywhere it appears (§18).
 */

import { engineeringConfig, STEEL_DENSITY } from '../config/engineeringConfig'
import { millConfig } from '../config/millConfig'
import type { MachineState } from '../types/machine'

export interface UtilisationSet {
  /** Roll separating force as % of the mill's maximum. */
  force: number
  /** Main drive torque as % of rating. */
  torque: number
  /** Main drive current as % of rating. */
  current: number
  /** Main drive power as % of rating. */
  power: number
  /** Mill speed as % of maximum. */
  speed: number
}

export function calculateUtilisation(state: MachineState): UtilisationSet {
  const { motorLimits, speedLimits } = engineeringConfig
  return {
    force: (state.rollingForce.actual / millConfig.ratings.maxRollingForce) * 100,
    torque: state.drive.torquePercentage,
    current: (state.drive.current / motorLimits.currentMax) * 100,
    power: (state.drive.power / motorLimits.powerMax) * 100,
    speed: (state.speed.actual / speedLimits.max) * 100,
  }
}

export interface ConsistencyChecks {
  /**
   * Mass-flow closure, % (§8.1). A healthy mill closes to well under 1%.
   * Surfaced as a diagnostic, never used to silently correct a value.
   */
  massFlowErrorPct: number
  massFlowOk: boolean
  /** Gaugemeter residual |h - (S0 + F/M)|, µm. */
  gaugemeterResidualUm: number
  gaugemeterOk: boolean
  /** Delivered thickness inside the ±tolerance target. */
  thicknessInTolerance: boolean
  /** Force within the schedule's prediction band. */
  forceVsScheduleErrorPct: number
}

/**
 * The three internal-consistency checks a commissioning engineer would run
 * against a twin (§18/§19). Exposed on screen so the model can be challenged
 * rather than trusted.
 */
export function calculateConsistency(state: MachineState): ConsistencyChecks {
  const referenceForce = state.rollingForce.reference
  const forceVsScheduleErrorPct =
    referenceForce > 0
      ? ((state.rollingForce.actual - referenceForce) / referenceForce) * 100
      : 0

  return {
    massFlowErrorPct: state.diagnostics.massFlowErrorPct,
    massFlowOk: Math.abs(state.diagnostics.massFlowErrorPct) < 1,
    gaugemeterResidualUm: state.diagnostics.gaugemeterResidualUm,
    gaugemeterOk: state.diagnostics.gaugemeterResidualUm < 1,
    thicknessInTolerance:
      Math.abs(state.thickness.deviation) <= engineeringConfig.thicknessTolerance,
    forceVsScheduleErrorPct,
  }
}

/**
 * Specific energy consumed by the current pass, kWh/t.
 *
 *   e = P / (ṁ)  with  ṁ = h·w·v·ρ
 *
 * A standard rolling-mill efficiency indicator. Returns 0 when the mill is
 * stopped rather than dividing by zero.
 */
export function specificEnergy(state: MachineState): number {
  const thicknessM = state.thickness.actual / 1000
  const widthM = state.coil.width / 1000
  const speedMps = state.speed.actual / 60
  // t/h = m³/s × kg/m³ × 3600 / 1000
  const throughputTph = thicknessM * widthM * speedMps * STEEL_DENSITY * 3.6
  if (throughputTph <= 0.001) return 0
  return state.drive.power / throughputTph
}

/** Production rate at the current speed and section, t/h. */
export function throughput(state: MachineState): number {
  const thicknessM = state.thickness.actual / 1000
  const widthM = state.coil.width / 1000
  const speedMps = state.speed.actual / 60
  return thicknessM * widthM * speedMps * STEEL_DENSITY * 3.6
}

/** Estimated time to finish the current pass, seconds. */
export function passTimeRemaining(state: MachineState): number | null {
  if (state.speed.actual <= 0.1) return null
  const entrySpeedMpm =
    state.thickness.entry > 0
      ? (state.speed.actual * state.thickness.actual) / state.thickness.entry
      : state.speed.actual
  return (state.coil.remainingLength / entrySpeedMpm) * 60
}

/** Format seconds as mm:ss for the pass countdown. */
export function formatDuration(seconds: number | null): string {
  if (seconds === null || !Number.isFinite(seconds)) return '—'
  const m = Math.floor(seconds / 60)
  const s = Math.floor(seconds % 60)
  return `${m}:${s.toString().padStart(2, '0')}`
}

/** Format seconds as hh:mm:ss for day totals such as rolling time. */
export function formatHMS(seconds: number | null): string {
  if (seconds === null || !Number.isFinite(seconds) || seconds < 0) return '—'
  const h = Math.floor(seconds / 3600)
  const m = Math.floor((seconds % 3600) / 60)
  const s = Math.floor(seconds % 60)
  return `${h}:${m.toString().padStart(2, '0')}:${s.toString().padStart(2, '0')}`
}

File: dotnet/src/Crm04.Domain/Machine/RollingEngine.cs
using System.Globalization;
using Crm04.Domain.Configuration;
using Crm04.Domain.Types;

namespace Crm04.Domain.Machine;

/// <summary>
/// Derived engineering indicators. Port of the pure helpers in <c>src/machine/rollingEngine.ts</c>.
///
/// EVERY VALUE HERE IS DERIVED FROM VALUES ALREADY ON SCREEN, and none of them is a tag. That is
/// why the UI renders them through <c>DerivedRow</c> in the dimmer treatment rather than through
/// <c>ValueReadout</c>: throughput is not a measurement, it is arithmetic on a thickness, a width
/// and a speed, and presenting it with the same weight as an instrument reading would overstate
/// what the mill actually told us.
///
/// They live in the domain, not the UI, so the number is computed once on the server and
/// transported - which keeps the §18 rule that a value on screen has exactly one source.
/// </summary>
public static class RollingEngine
{
    /// <summary>
    /// Production rate at the current speed and section, t/h.
    ///
    /// t/h = m³/s × kg/m³ × 3600 / 1000, which collapses to the 3.6 factor below.
    /// </summary>
    public static double Throughput(MachineState state)
    {
        var thicknessM = state.Thickness.Actual / 1000d;
        var widthM = state.Coil.Width / 1000d;
        var speedMps = state.Speed.Actual / 60d;
        return thicknessM * widthM * speedMps * EngineeringConfig.SteelDensity * 3.6d;
    }

    /// <summary>
    /// Specific energy, kWh/t. Zero at standstill rather than an infinity: dividing drive power
    /// by a throughput of nothing is not a large number, it is an undefined one.
    /// </summary>
    public static double SpecificEnergy(MachineState state)
    {
        var tph = Throughput(state);
        if (tph <= 0.001d) return 0d;
        return state.Drive.Power / tph;
    }

    /// <summary>
    /// Estimated time to finish the current pass, seconds, or null at standstill.
    ///
    /// Uses the ENTRY speed, because it is entry-side strip that is being consumed from the payoff
    /// reel. Using the exit speed would under-estimate the time by the reduction ratio.
    /// </summary>
    public static double? PassTimeRemaining(MachineState state)
    {
        if (state.Speed.Actual <= 0.1d) return null;

        var entrySpeedMpm = state.Thickness.Entry > 0d
            ? state.Speed.Actual * state.Thickness.Actual / state.Thickness.Entry
            : state.Speed.Actual;

        return state.Coil.RemainingLength / entrySpeedMpm * 60d;
    }

    /// <summary>Format seconds as m:ss for the pass countdown. An em dash when unknown.</summary>
    public static string FormatDuration(double? seconds)
    {
        if (seconds is not { } s || !double.IsFinite(s)) return "—";
        var m = (int)Math.Floor(s / 60d);
        var sec = (int)Math.Floor(s % 60d);
        return $"{m}:{sec.ToString("00", CultureInfo.InvariantCulture)}";
    }

    /// <summary>Format seconds as h:mm:ss for the rolling-time total.</summary>
    public static string FormatHms(double? seconds)
    {
        if (seconds is not { } s || !double.IsFinite(s) || s < 0d) return "—";
        var h = (int)Math.Floor(s / 3600d);
        var m = (int)Math.Floor(s % 3600d / 60d);
        var sec = (int)Math.Floor(s % 60d);
        return $"{h}:{m.ToString("00", CultureInfo.InvariantCulture)}:{sec.ToString("00", CultureInfo.InvariantCulture)}";
    }
}
